syntax = "proto3";

package kiorga.datamodel;

import "google/protobuf/timestamp.proto";

// Enum zur Definition des Bearbeitungsstatus einer Aufgabe.
enum TaskStatus {
  TASK_STATUS_UNSPECIFIED = 0;
  TASK_STATUS_PENDING = 1;
  TASK_STATUS_IN_PROGRESS = 2;
  TASK_STATUS_COMPLETED = 3;
  TASK_STATUS_FAILED = 4;
}

// Enum zur Definition der Priorität einer Aufgabe.
enum TaskPriority {
  TASK_PRIORITY_UNSPECIFIED = 0;
  URGENT = 1;
  HIGH = 2;
  MEDIUM = 3;
  LOW = 4;
  OPTIONAL = 5;
}

// Das zentrale Objekt zur Steuerung und Verfolgung von Arbeit in der KI-Organisation.
message Task {
  // Die global eindeutige ID der Aufgabe (z.B. eine UUID).
  string task_id = 1;

  // Kurzer, menschenlesbarer Titel der Aufgabe, z.B. für Listenansichten.
  string title = 2;

  // Detaillierte Beschreibung der Arbeitsanweisung für den ausführenden Agenten.
  string description = 3;

  // Der aktuelle Bearbeitungsstatus der Aufgabe.
  TaskStatus status = 4;

  // Die Priorität der Aufgabe zur Steuerung der Abarbeitungsreihenfolge.
  TaskPriority priority = 5;

  // Die ID des Agenten, der für die Ausführung der Aufgabe verantwortlich ist.
  string assigned_to_agent_id = 6;

  // Die ID des Agenten, der die Aufgabe ursprünglich erstellt hat.
  string creator_agent_id = 7;

  // Zeitstempel der Erstellung der Aufgabe.
  google.protobuf.Timestamp created_at = 8;

  // Zeitstempel der letzten Aktualisierung. Wird bei jeder Änderung aktualisiert.
  google.protobuf.Timestamp updated_at = 9;

  // Eine Liste von 'task_id's, die abgeschlossen sein müssen, bevor diese Aufgabe starten kann.
  repeated string dependencies = 10;

  // Schlüssel-Wert-Paare für Verweise auf Input-Daten (z.B. Pfade in Cloud Storage).
  map<string, string> input_data_references = 11;

  // Schlüssel-Wert-Paare für Verweise auf Output-Daten.
  map<string, string> output_data_references = 12;

  // Textuelle Beschreibung der Erfolgskriterien ("Definition of Done").
  string success_criteria_metrics = 13;

  // Optional: Die ID der übergeordneten Aufgabe zur Abbildung von Hierarchien.
  string parent_task_id = 14;

  // Optional: Das Fälligkeitsdatum für die Aufgabe, um die Priorisierung zu unterstützen.
  google.protobuf.Timestamp due_date = 15;

  // Wenn 'true', erfordert die Aufgabe menschliche Freigaben an definierten Punkten.
  bool co_pilot_mode = 16;
}